### DISCUSSION TITLE
Some system architecture questions about Knative eventing and Dapr

### DISCUSSION THREAD

**Discussion by JasonChen86899** [OpenFunction/OpenFunction#430] [Q&A] [answered]
Hi openfunction team member. I have some following questions of the system architecture

1.  Knative eventing is like openfunction event framework but also has autoscaling like KEDA. we can use it

2.  IMO Dapr is an appliction runtime, it can proxy many componts and it can also do async eg: pub/sub, but it may be the same with event framework in async. when using Knative eventing or openfunction event framework, dapr is unnecessary for async and eventing framework can do it. so just use it as an application runtime

3. I seen WsamEdge that is good! so can run wsam+dapr which can do application runtime

Thx

**Comment by benjaminhuo** [comment]
> * Knative eventing is like openfunction event framework but also has autoscaling like KEDA. we can use it
> *

Knative eventing is a bit complex to use and there're too many crds(one crd for each event source). Dapr input binding and pubsub can replace knative eventing without problem

**Reply by JasonChen86899** [reply]
yeah. openfunction event framework + darp pubsub can do same with knative eventing. But as you say Knative eventing is complex that means complex in eventing system. except of event source and trigger, sink,  filter, channel, flow, deadletter so on are also necessary in an production eventing system.  maybe I want to join a chat group of Openfunction for detail discussion😄

**Reply by benjaminhuo** [reply]
> yeah. openfunction event framework + darp pubsub can do same with knative eventing. But as you say Knative eventing is complex that means complex in eventing system. except of event source and trigger, sink, filter, channel, flow, deadletter so on are also necessary in an production eventing system. maybe I want to join a chat group of Openfunction for detail discussion😄

Sorry for my delay.
We use wechat group and discord frequently.
You can find methods to join these two in https://github.com/OpenFunction/community#contact-us 
@JasonChen86899

**Comment by benjaminhuo** [comment, marked as answer]
> 3\. I seen WsamEdge that is good! so can run wsam+dapr which can do application runtime

Dapr rust sdk is not as mature as other sdks, we can wait a bit longer to implement that
